packages/wdio-browserstack-service/src/Percy/Percy-Handler.ts

```diff
@@ -0,0 +1,184 @@
+import type { Capabilities } from '@wdio/types'
+import type { BeforeCommandArgs, AfterCommandArgs } from '@wdio/reporter'
+
+import {
+    o11yClassErrorHandler,
+    sleep
+} from '../util.js'
+import PercyCaptureMap from './PercyCaptureMap.js'
+
+import * as PercySDK from './PercySDK.js'
+import { PercyLogger } from './PercyLogger.js'
+
+import { PERCY_DOM_CHANGING_COMMANDS_ENDPOINTS, CAPTURE_MODES } from '../constants.js'
+
+class _PercyHandler {
+    private _testMetadata: { [key: string]: any } = {}
+    private sessionName?: string
+    private _isAppAutomate?: boolean
+    private isPercyCleanupProcessingUnderway?: boolean = false
+    private _percyScreenshotCounter: any = 0
+    private percyDeferredScreenshots: any = []
+    private percyScreenshotInterval: any = null
+
+    constructor (
+        private _percyAutoCaptureMode: string | undefined,
+        private _browser: WebdriverIO.Browser | WebdriverIO.MultiRemoteBrowser,
+        private _capabilities: Capabilities.RemoteCapability,
+        isAppAutomate?: boolean,
+        private _framework?: string
+    ) {
+        this._isAppAutomate = isAppAutomate
+        if (!_percyAutoCaptureMode || !CAPTURE_MODES.includes(_percyAutoCaptureMode as string)) {
+            this._percyAutoCaptureMode = 'auto'
+        }
+    }
+
+    _setSessionName(name: string) {
+        this.sessionName = name
+    }
+
+    async teardown () {
+        await new Promise<void>((resolve) => {
+            setInterval(() => {
+                if (this._percyScreenshotCounter === 0) {
+                    resolve()
+                }
+            }, 1000)
+        })
+    }
+
+    async percyAutoCapture(eventName: string | null, sessionName: string | null) {
+        try {
+            if (eventName) {
+                if (!sessionName) {
+                    /* Service doesn't wait for handling of browser commands so the below counter is used in teardown method to delay service exit */
+                    this._percyScreenshotCounter += 1
+                }
+
+                (this._browser.percyCaptureMap as PercyCaptureMap).increment(sessionName ? sessionName : (this.sessionName as string), eventName)
```

Comment:
Is there any reason for attaching this object to the `browser`? If this is not used or supposed to be used by the user it shouldn't be attached to the browser object.

Furthermore since you have extended the WebdriverIO types in [here](packages/wdio-browserstack-service/src/@types/bstack-service-types.d.ts) there should be no need for casting. Please make sure the type is properly respected by TS.

Reply:
Fixed the approach here, not attaching percyCaptureMap object to browser anymore.